File: 백준/Gold/15486. 퇴사 2/퇴사 2.py
n = int(input())
t = [0]
p = [0]
for _ in range(n):
    a,b = map(int,input().split())
    t.append(a)
    p.append(b)
    
dp = [0] * (n + 2)

for i in range(1, n + 1):
    # 상담을 할수 있는경우
    if i + t[i] - 1 <= n:
        dp[i + t[i] - 1] = max(dp[i + t[i] - 1], dp[i - 1] + p[i])
    # 상담을 하지 않고 다음 날로 넘기는 경우
    dp[i] = max(dp[i], dp[i - 1])
print(dp[n])

File: 백준/Gold/15486. 퇴사 2/퇴사 2.js
const input = require("fs").readFileSync("/dev/stdin", "utf-8").trim().split("\n");

const n = Number(input[0]);
const t = [0];
const p = [0];

for (let i = 1; i <= n; i++) {
	const [a, b] = input[i].trim().split(" ").map(Number);
	t.push(a);
	p.push(b);
}

const dp = Array(n + 1).fill(0);

for (let i = 1; i < n + 1; i++) {
	if (i + t[i] - 1 <= n) {
		dp[i + t[i] - 1] = Math.max(dp[i + t[i] - 1], dp[i - 1] + p[i]);
	}
	dp[i] = Math.max(dp[i], dp[i - 1]);
}
console.log(dp[n]);
